### PR TITLE
Proposed fix for #3987

### PR DESCRIPTION
**Pull request by larseggert**
In the interest of helping move things along, here's a proposal for #3987.

### CODE REVIEW

**Review by ianswett — COMMENTED** [review]
I think we're using PTO in two cases because we want to allow time for the PTO to fire twice, but I could be wrong?

**Comment by larseggert** [comment]
The issue that this is intending to address isn't about "why use a multiple of PTO for the timer" but "why use PTO, which includes max ACK delay, and not e.g. derive from SRTT". At least that is how I understood it...

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2693,6 +2698,10 @@ ensures that connections are not closed after new activity is initiated.
 
 To avoid excessively small idle timeout periods, endpoints MUST increase the
 idle timeout period to be at least three times the current Probe Timeout (PTO).
+As defined in {{QUIC-RECOVERY}}, the PTO includes the peer's maximum expected
+acknowledgement delay. This is not strictly necessary for the purposes of the
+idle timeout, but defining it in this way is convenient, since implementations
+need to maintain a PTO value anyway.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2704]:
```suggestion

Note:

: Using the PTO value here implicitly includes the peer's maximum expected
acknowledgement delay in the idle timeout period; see Section 6.2.1
of {{QUIC-RECOVERY}}. This ensures that the endpoint accounts for this
known peer delay before declaring the connection as idle and closing it.
```

Comment (janaiyengar) [reply draft-ietf-quic-transport.md:2704]: This one is deliberate -- we want to account for max ack delay.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2760,7 +2769,10 @@ application-supplied error code will be used to signal closure to the peer.
 The closing and draining connection states exist to ensure that connections
 close cleanly and that delayed or reordered packets are properly discarded.
 These states SHOULD persist for at least three times the current Probe Timeout
-(PTO) interval as defined in {{QUIC-RECOVERY}}.
+(PTO) interval as defined in {{QUIC-RECOVERY}}. Note that the PTO includes the
+peer's maximum expected acknowledgement delay. While not strictly necessary for
+this purpose, it is convenient, since implementations need to maintain a PTO
+value anyway.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2775]:
```suggestion
(PTO) interval as defined in {{QUIC-RECOVERY}}.

Note:

: Using the PTO value here implicitly includes the peer's maximum expected
acknowledgement delay in the closing and draining periods; see Section 6.2.1
of {{QUIC-RECOVERY}}. This ensures that the endpoint takes this peer delay
into account when waiting to discard subsequent packets sent by the peer.
```

**Comment by ianswett** [comment]
@larseggert in some cases, the reason for using PTO is to ensure there's time to PTO retransmit, so I don't think this change currently captures that reason in this cases.

I agree that's the basis of the issue as it stands, but I think the better question is why use PTO in different cases.  I think the max_ack_delay is not particularly interesting in any of these cases.

**Review by ianswett — CHANGES_REQUESTED** [review]
I don't think we need all this extra text on max_ack_delay in these cases, even if that's how the issue was filed.

**Comment by larseggert** [comment]
I'm completely fine with not making changes. @Ralith?

**Comment by Ralith** [comment]
Maybe just reduce it to the first case? My goal with the issue was make it explicit when we use PTO, including max_ack_delay, for the sake of simplicity even though max_ack_delay isn't actually pertinent. If it serves a legitimate purpose for idle/close timers, then there's less need to justify its involvement, because that involvement is less likely to produce confusion and give the impression of a spec bug.

**Review by janaiyengar — APPROVED** [review]

**Review by martinthomson — COMMENTED** [review]
I think that the document would be fine without any of these.  Certainly the latter two are a distraction.

**Comment by larseggert** [comment]
OK, I rev'ed the PR to only include the first note. Do folks think this is OK, do we want changes, or do we want to drop if altogether?

**Comment by ianswett** [comment]
I sent out an alternate PR: https://github.com/quicwg/base-drafts/pull/4084

**Comment by janaiyengar** [comment]
@ianswett pointed out that the point here is not so much about including max_ack_delay, it is about using PTO values. Thinking about this, I agree. The idea here is to basically allow for the sender to retransmit the probe multiple times, and that happens on PTOs. Basically, I think what needs to be written is an explanation of the expression. My suggested explanation and text:

`6 x InitialRTT` ensures that the period is large enough for sending multiple PATH_CHALLENGE frames on initial timeout (with exponential backoff). `3 x PTO` ensures that _if the current path PTO is longer than `2 x InitialRTT`_  then we allow for the new path to be as long. If we want to clarify something here, that would be the point to clarify.

**Comment by janaiyengar** [comment]
Uh, sorry, @ianswett : I posted my message and then saw your message (and PR) above. Looking now.

**Pull request closed without merging**